Don't link transitive static carthage libraries

akkyie opened the pull request:
The current version of XcodeGen tends to link all static carthage libraries transitively.
That means it links static dependencies of an App target to its test target. 
https://github.com/yonaskolb/XcodeGen/blob/aaae84f883c31dc0a4050f56942ee99c1b109135/Sources/XcodeGenKit/PBXProjGenerator.swift#L910-L925

Images below illustrate this behavior, where SwiftyJSON is a static library linked to the app target, and swift-tagged is another one linked directly to the test target.

|App|Test|
|---|----|
|<img width="300" alt="Screen Shot 2020-04-30 at 13 34 48" src="https://user-images.githubusercontent.com/1528813/80672958-fecb3f00-8ae8-11ea-827f-bba93a71841a.png">|<img width="300" alt="Screen Shot 2020-04-30 at 13 35 00" src="https://user-images.githubusercontent.com/1528813/80672951-fb37b800-8ae8-11ea-83c5-8ca7e92ceeed.png">|

I believe this is not how we expect it to work because SwiftyJSON binary is embedded into both app and test executable, it causes symbol collision when you run test.

I changed PBXProjGenerator to link static libraries only what the top-level target depends, but not transitive ones. The count of results of `CarthageDependencyResolver.dependencies(for:)` won't change, as I'd rather want to keep its completeness.

Dependencies of the test target after this change:
<img width="300" alt="Screen Shot 2020-04-30 at 13 35 51" src="https://user-images.githubusercontent.com/1528813/80673838-46eb6100-8aeb-11ea-977c-fc34a088e8e1.png">

I just added SwiftyJSON and swift-tagged statically to the fixture project in 7e65bd7 first, to see the actual diff for the code changes in 1da2133 (and have no opinion about the choice of libraries; just lightweight ones.)

brentleyjones commented on Tests/Fixtures/TestProject/Cartfile:
```diff
@@ -1 +1,3 @@
\ No newline at end of file
+github "antitypical/Result"
+github "SwiftyJSON/SwiftyJSON"
+github "pointfreeco/swift-nonempty"
```
Nit: Can you add a trailing newline?

akkyie commented:
@brentleyjones Thanks! Added newline and merged master. Found fixture generation becoming indeterministic, also fixed.

> Do we have similar issues with transitive static library Target or SwiftPM dependencies?

I now tried both cases and they doesn't look to have problems. The app in the fixture project already has a static dependency but it's test is fine, and checked the SPM case on my local.
Should I also add it to the fixture?

brentleyjones commented:
If you could strengthen the fixture that would be great, thanks.

akkyie commented:
@brentleyjones Added SPM fixture changes and test fixes 👍  79621eb 2dde630

brentleyjones commented:
Looks like we have some test failures.

akkyie commented:
Oops, it succeeded with the empty source directory only on my local. Added a source.

akkyie commented:
Sorry again, fixing a typo in the spec.

akkyie commented:
Looks fine now?
